pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/loader/IndexLoadingConfig.java

```diff
@@ -233,8 +235,14 @@ private void extractFromTableConfigAndSchema(TableConfig tableConfig, @Nullable
     }
 
     List<String> onHeapDictionaryColumns = indexingConfig.getOnHeapDictionaryColumns();
+    Map<String, OnHeapDictionaryConfig> onHeapDictionaryConfigMap = indexingConfig.getOnHeapDictionaryConfigs();
     if (onHeapDictionaryColumns != null) {
-      _onHeapDictionaryColumns.addAll(onHeapDictionaryColumns);
+      for (String col : onHeapDictionaryColumns) {
+        _onHeapDictionaryColumns.add(col);
+        if (onHeapDictionaryConfigMap != null && onHeapDictionaryConfigMap.containsKey(col)) {
+          _onHeapDictionaryConfigs.put(col, onHeapDictionaryConfigMap.get(col));
```

Comment:
As said in IndexingConfig comment, I recommend to do not support this feature using the old syntax. By doing so we won't need to modify this part of the code.